### ISSUE TITLE
Compiling on Linux and using with Wine like a n00b?

### ISSUE THREAD

**Issue by mrmatteastwood** [kcat/dsoal#56]
Hey @kcat I'm hoping you can help me with a bit of an unusual request. I'm trying to compile the latest version of DSOAL on my Linux Mint 20.2 machine and then to use it with old games running on Wine or Proton. Even though I've been using Linux since 2009, I'm a beginner when it comes to compiling stuff and also I have a ton yet to learn about Wine.

I'm hoping you're willing and able to walk me through this process.

**So I got this far:**

- Installed CMake-GUI and mingw-w64 (as suggested in issue #27)
- InCMake-GUI, set source and build dir, then clicked "Configure" and then "Generate"
- In terminal, went to build directory and typed "make"

Then this happens:

```
Scanning dependencies of target dsound
[ 10%] Building C object CMakeFiles/dsound.dir/buffer.c.o
/home/thebrightside/zzLocal/Test/dsoal-master/buffer.c:27:10: fatal error: windows.h: No such file or directory
   27 | #include "windows.h"
      |          ^~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/dsound.dir/build.make:63: CMakeFiles/dsound.dir/buffer.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/dsound.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```

This is where I'm stuck.

Also, I'm really hazy on where to go if I even got the thing to compile. I read online that all you need is to put dsound.dll into the same directory as a game's executable, but then I also read that I also need another DLL called, "dsoal-aldrv.dll" but I'm not sure where to get that from.

Apparently, you also have another project called, "openal-soft"? Is that where I get the dsoal-aldrv.dll? Or if it's supposed to come from the Windows directory, where would I get it if I run Wine?

Do I even have a realistic chance to restore surround sound in old games using DSOAL if I'm only on Wine and not on real Windows?

================================================

I realize this is a bit of an unusual endeavour and I'm sure I get stuck at kinda stupid points, so perhaps you can nudge me in the right directions. I'm happy to learn and promise to not badger you too much ;-)

If, somehow, this ends up working, I plan to post tutorials on forums and make a YouTube video so that other surround gaming and Linux fans can benefit.

**Comment by mirh** [comment]
Aldrv is openal-soft, yes. 
Also, wine is already using openal for at least some of its audio IIRC. 
Also also, the proper way for you to get this fixed there would be them directly supporting the thing (if just so it wasn't they seem to have [forgotten](https://bugs.winehq.org/show_bug.cgi?id=40898#c8)). 

Anyway, you should just need to install the mingw-w64-headers to pull this off.

**Comment by kcat** [comment]
To use it with Wine, you essentially just need to cross-compile it for Windows, and use the resulting DLL with the app in Wine. You'll need a toolchain file to have CMake cross-compile the project with MinGW, one like this: https://github.com/kcat/openal-soft/blob/1.22.0/XCompile.txt

You'll also need [OpenAL Soft](https://github.com/kcat/openal-soft) as a Windows DLL too, named "dsoal-aldrv.dll" for DSOAL to recognize it.

**Comment by mrmatteastwood** [comment]
Hey @kcat thanks for your response :-) I looked into this and found some more info about cross-compiling and toolchain files, but even after reading up on it, I'm not getting any further. I've been meaning to learn programming, but I never found the time :-(

I saved your XCompile.txt in the root directory and then this happens:

```
thebrightside@CHAINSAWBOI:~/zzLocal/Test/dsoal-master/build$ cmake -DCMAKE_TOOLCHAIN_FILE=../XCompile.txt ..
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_TOOLCHAIN_FILE


-- Build files have been written to: /home/thebrightside/zzLocal/Test/dsoal-master/build
```

I now have a bunch of files in that build folder, but I don't know what do to with them and that warning above doesn't fill me with confidence either. When I type "make", I get the same outcome as in my OP.

**Comment by mrmatteastwood** [comment]
Follow-up: I think I understand what I did wrong. I didn't delete the already-created build files. Now, I get this:

```
cmake -DCMAKE_TOOLCHAIN_FILE=../XCompile.txt ..
-- The C compiler identification is unknown
CMake Error at CMakeLists.txt:3 (project):
  The CMAKE_C_COMPILER:

    -gcc

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.


-- Configuring incomplete, errors occurred!
See also "/home/thebrightside/zzLocal/Test/dsoal-master/build/CMakeFiles/CMakeOutput.log".
See also "/home/thebrightside/zzLocal/Test/dsoal-master/build/CMakeFiles/CMakeError.log".
```

I'm guessing I need to edit some lines in your XCompile.txt. Unfortunately, I don't understand what needs to be edited.

**Comment by kcat** [comment]
As mentioned in the file, you also need to specify something like `-DHOST=i686-w64-mingw32` with it (for 32-bit MinGW, or `-DHOST=x86_64-w64-mingw32` for 64-bit MinGW).

**Comment by mrmatteastwood** [comment]
Nice, that did it. I was able to compile it using the 32-bit version.
Next, I downloaded openal-soft-1.22.0-bin.zip from the openal-soft website, then extracted soft_oal.dll from its bin/Win32 folder, renamed it to dsoal-aldrv.dll and placed both files in the root directory of Prey (2006).

I then launched Prey from Steam (I previously deleted any existing Proton data that the game had). Right now, when I try to select "OpenAL" as the sound system in the game menu, it checks for the fraction of a second and then goes back to "Default". In the speaker settings, I can still select "Surround", but the game definitely runs in stereo only. I'll try a few more things.

**Comment by mirh** [comment]
You have to set dlls to native for wine to use what you provide. And I'd avoid proton if I were you.

**Comment by kcat** [comment]
> Nice, that did it. I was able to compile it using the 32-bit version. When attempting 64-bit, I got an error: "...is not a full path and was not found in the PATH."

You need to install the packages for 64-bit MinGW (on my Debian system, that's `mingw-w64-x86-64-dev`, although there is the `mingw-w64` package which installs the stuff for both 32-bit and 64-bit MinGW).

> Next, I downloaded openal-soft-1.22.0-bin.zip from the openal-soft website, then extracted soft_oal.dll from its bin/Win32 folder, renamed that file to dsoal-aldrv.dll and placed both files in the root directory of Prey (2006).
>
> I then launched Prey from Steam (I previously deleted any existing Proton data that the game had). Right now, when I try to select "OpenAL" as the sound system in the game menu, it checks for the fraction of a second and then goes back to "Default".

If the game uses OpenAL natively, you don't need DSOAL. Prey (2006) uses the idTech 5 engine which uses OpenAL natively, so DSOAL won't do much (unless it's using the Generic Hardware wrapper, which maps OpenAL -> DSound (where DSOAL then maps DSound -> OpenAL)). You're better off using OpenAL Soft directly by renaming soft_oal.dll to OpenAL32.dll and placing it in the game directory next to the exe.

> And I'd avoid proton if I were you.

Why?

**Comment by mirh** [comment]
It's more of a bother to customize (winetricks is everywhere, not so much protontricks) and it's shakier to track down issues (let alone if they have some special hack like for media codecs).

**Comment by mrmatteastwood** [comment]
Hell yeah, Prey works!! Thanks for your help, @kcat. You were of course right, all that was necessary was OAL-Soft. Once I placed the 32-bit DLL in the Prey root directory, with the file named OpenAL32.dll, surround sound worked out of the box. No additional config was necessary in Proton at all.

I updated [Prey's PC Gaming Wiki page](https://www.pcgamingwiki.com/wiki/Prey#Sound_doesn.27t_properly_work_on_Linux) with the info.

I'm installing F.E.A.R. next and will try DSOAL with that as soon as I find time. Will report back.

**Comment by mirh** [comment]
None of that is really an "issue", let alone specific to linux or windows 10. It's more like working as intended with respect to the openal design (which I thought wine shipped by default btw?)
See for example the mass effect page for how it should be treated (even though I reckon I never came back to polish it with the latest findings)

**Comment by mrmatteastwood** [comment]
Hey @mirh yeh you have a point, what we're discussing here isn't really a "bug" and I understand I'm kind of mis-using the issue system for this. I appreciate that you guys are indulging me.

For me, this thread is incredibly helpful to understand more about the different sound systems that games use, how they work and what to do if they don't. Even though I've had 5.1/7.1 home theatre systems hooked up to my PC since 2009, I was always used to it "just working" in Windows, except for a few cases that usually had guides online: simple stuff like "get IndirectSound, put DLL here, bam works."

With Windows 11 being incompatible with a lot of powerful gaming PCs (and its increasingly invasive shittiness), I'm pursuing 2 goals: 1. be Windows-free by 2025, when Win 10 reaches EoL, and b) give back to the expected influx of Linux newbies who want to play in surround.

Folks like @kcat are doing tremendous work in surround sound preservation, and I expect with the aforementioned Win11 situation and the popularity of the Steam Deck, we'll see more people coming to Linux who are also surround audio enthusiasts. They might wonder how to restore surround functionality in Prince of Persia 2008, Serious Sam 2, F.E.A.R., Mirror's Edge and other titles that (at present) don't work out of the box.

So anyway, long story short, I've learned a whole lot in the past month or so, exploring gaming on Linux, and I've been taking a whole bunch of step-by-step notes for myself. Whenever I find time, I plan to turn those into newbie-friendly guides and perhaps put them on YT, the gamingonlinux forum or some such.

**Comment by mrmatteastwood** [comment]
Oh, with all the ranting, I forgot to ask: would you mind pointing me to that Mass Effect page you mentioned?

**Comment by mrmatteastwood** [comment]
Thanks for the links! More reading material. Will dig into this. True, Win 11's BS isn't new. What's new is the insane amount of games that now run on Linux ;-)

**Comment by mrmatteastwood** [comment]
Finally had a chance to try this with F.E.A.R., and unfortunately, it doesn't work.

I downloaded the installer files from GOG, then installed F.E.A.R. using Lutris. Set Wine to Windows XP using winecfg for the game to run. First run was without any tweaks and the game was in stereo, all surround options (HW Mixing, EAX 2.0 and EAX Advanced HD) greyed out in sound menu.

Second run was with 32-bit dsound.dll and dsoal-aldrv.dll in the same folder as FEAR.exe. Same result as above.

For shits and giggles, I removed dsound.dll and dsoal-aldrv.dll again, installed xact, dsound and vcrun2010 using Winetricks, then tried IndirectSound's dsound.dll. Same result as above.

Then, put DSOAL files back (now that xact, dsound and vcrun2010 were installed). Same result as above.

FWIW, this is what I get in the terminal (from the fourth run):

```
thebrightside@CHAINSAWBOI:~$ lutris
2022-06-24 13:24:56,681: Starting Lutris 0.5.10.1
2022-06-24 13:24:56,705: No folder at /home/thebrightside/.local/share/lutris/runners/retroarch/
2022-06-24 13:24:56,798: Using NVIDIA drivers 510.73.05 for x86_64
2022-06-24 13:24:56,798: GPU: NVIDIA GeForce GTX 1080 Ti
2022-06-24 13:24:56,798: GPU: 10DE:1B06 3842:6696 (nvidia drivers)
2022-06-24 13:24:56,875: Startup complete
Exception ignored in: <function InputDevice.__del__ at 0x7f2067dd10d0>
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/evdev/device.py", line 159, in __del__
    self.close()
  File "/usr/lib/python3/dist-packages/evdev/device.py", line 304, in close
    super().close()
  File "/usr/lib/python3/dist-packages/evdev/eventio_async.py", line 54, in close
    loop = asyncio.get_event_loop()
  File "/usr/lib/python3.8/asyncio/events.py", line 639, in get_event_loop
    raise RuntimeError('There is no current event loop in thread %r.'
RuntimeError: There is no current event loop in thread 'Thread-2'.
Started initial process 117322 from gamemoderun /home/thebrightside/.local/share/lutris/runners/wine/lutris-fshack-7.2-x86_64/bin/wine /home/thebrightside/zzLocal/Appdata/Lutris/gog/fear/drive_c/GOG Games/F.E.A.R. Platinum Collection/FEAR.exe
Start monitoring process.
ERROR: ld.so: object '/usr/$LIB/libgamemodeauto.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/usr/$LIB/libgamemodeauto.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
esync: up and running.
ERROR: ld.so: object '/usr/$LIB/libgamemodeauto.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
Note: Illegal Audio-MPEG-Header 0x91500bf1 at offset 47955.
Note: Trying to resync...
Note: Skipped 382 bytes in input.
Note: Illegal Audio-MPEG-Header 0x91500bf1 at offset 47955.
Note: Trying to resync...
Note: Skipped 382 bytes in input.
Note: Illegal Audio-MPEG-Header 0x5ab36954 at offset 47955.
Note: Trying to resync...
Note: Skipped 68 bytes in input.
Note: Illegal Audio-MPEG-Header 0x5ab36954 at offset 47955.
Note: Trying to resync...
Note: Skipped 68 bytes in input.
Initial process has exited (return code: 0)
Monitored process exited.
Exit with return code 0
2022-06-24 13:28:48,802: Game still running (state: running)
2022-06-24 13:28:48,803: Stopping F.E.A.R. (wine)
2022-06-24 13:28:48,821: Running post-exit command: /home/thebrightside/Appdata/Bash_Scripts/vsync-desktop-sb-3.sh
2022-06-24 13:28:49,025: Switching resolution of DVI-D-0 to 1280x1024
2022-06-24 13:28:49,238: Switching resolution of HDMI-0 to 1920x1080
2022-06-24 13:28:49,493: Switching resolution of DP-0 to 2560x1440
-> Red  1.000, Green  1.000, Blue  1.000
<- Red  1.000, Green  1.000, Blue  1.000
2022-06-24 13:28:58,407: Shutting down Lutris
```

Any idea what I might be doing wrong? Any suggestions will be very much appreciated :-) Thanks!

**Comment by mirh** [comment]
As I said a trainload of times, you have to take into account whether wine will load native or builtin
https://wiki.winehq.org/Winecfg#Libraries

**Comment by mrmatteastwood** [comment]
Ah, yes. The result stays unchanged whether I set openal32 to "native" or "builtin".

**Comment by mirh** [comment]
You are setting dsound.dll here...
Dsoal has nothing to do with system-wide openal.

**Comment by mrmatteastwood** [comment]
Thanks! I set dsound to both native and builtin, and in both cases, the sound options remain greyed out. I also tried setting Wine to Windows 7, to no avail. Same result.

In case the dll I compiled has something wrong with it, I also downloaded the old DSOAL from here: https://www.nexusmods.com/newvegas/mods/65094

Again, no dice.

**Comment by mirh** [comment]
Setting the dsoal log variable and seeing if anything comes out wouldn't hurt. 
https://old.reddit.com/r/wine_gaming/comments/lqtqpg/using_eax_20_on_wine_version_59/
https://old.reddit.com/r/GuildWars/comments/nj6b30/announcing_dsoalgw1_3d_positional_audio_and_eax/
They got it working here.

**Comment by mrmatteastwood** [comment]
Got some good news this morning! Tried my compiled DSOAL dll with Duke Nukem: Manhattan Project and got surround sound. So this means the file I compiled is definitely fine. I'm really excited about this!!

For F.E.A.R., I put the following lines into both my ~/.bashrc and /etc/profile:

export DSOAL_LOGLEVEL=3
export DSOAL_LOGFILE=1_dsoal_log.txt

I read that Wine takes these environment variables and uses them. However, no log file was written when I ran F.E.A.R. or DNMP.

Same result when I try this line in ~/.bashrc and /etc/profile:

export DSOAL_LOGFILE=/home/thebrightside/zzLocal/Gaming/1_dsoal_log.txt

**Comment by mrmatteastwood** [comment]
Oh, and I followed the steps from the reddit thread you linked to the letter. Thanks for these!

**Comment by mirh** [comment]
Sigh.. Maybe you should try to check with something more simple like rightmark, I don't know. 
Also https://wiki.winehq.org/Wine_User%27s_Guide#WINEDLLOVERRIDES.3DDLL_Overrides

**Comment by mrmatteastwood** [comment]
Tell ya what, I think we can close this issue, because my original objective was clearly attained. I was able to compile DSOAL and use it with Wine. I clearly couldn't have gotten nearly as far without your help, so again thanks for that.

F.E.A.R. being bitchy is a separate issue that I think I'll come back to when Linux Mint 21 is out. Until then, I have enough reading to do - I have yet to understand much of the links you shared. Plus I'm gonna see if some of what I learned can be useful for others.

Ta!
